shell/common/rasterizer.cc

```diff
@@ -802,6 +810,56 @@ std::optional<size_t> Rasterizer::GetResourceCacheMaxBytes() const {
   return std::nullopt;
 }
 
+void Rasterizer::MaybeSleepBeforeSubmit(
+    FrameTimingsRecorder& frame_timings_recorder) {
+  // A simple threshold. Can change to fancier approaches later.
+  static const int MAX_HISTORY = 10;
+  static const int HISTORY_THRESH = 2;
+  // a bit more than enough to allow errors
+  static const fml::TimeDelta SAFE_MARGIN =
+      fml::TimeDelta::FromMicroseconds(500);
+
+  // TODO should not assume 60FPS - fix it before PR merged
+  fml::TimeDelta FRAME_DURATION = fml::TimeDelta::FromMicroseconds(16777);
```

Comment:
This will not work.

When under heavy load, we have no way of knowing when we'll get rescheduled. This is very unsafe and will definitely lead to missed frames on lower end devices.

Reply:
(besides the fact that the vsync time is not necessarily correct, which you commented on already)

Reply:
To be clearer: when I say this will not work, I mean that it will not scale to representative devices that many Flutter users around the world use in real-world applications. I am certain it will work relatively well on higher end devices with small applications.

Reply:
> When under heavy load, we have no way of knowing when we'll get rescheduled. This is very unsafe and will definitely lead to missed frames on lower end devices.

Good insights! IMHO you mean `sleep()` will make us have to be rescheduled by OS later?

Then what about this simple fix: Do not `sleep`, but busy-wait. This will be almost equivalent to a slower rasterization.

> (besides the fact that the vsync time is not necessarily correct, which you commented on already)

Yes I just want a quick review before doing all the details in case the direction needs to be overhauled

Reply:
For example, suppose most rasterization ends in 15ms and this one ends in 10ms. Then we `while(true) noop();` for 5ms. This is as if we do some real rasterization work for 5ms, since both of them uses the CPU, and OS will treat the two cases the same.

Reply:
If we busy wait on this thread we'll be blocking rasterization work... If we sleep we have no way of knowing if the OS will actually reschedule us when we want (just that it probably won't be sooner than that).

Reply:
@dnfield Sorry do not quite get it...

> If we busy wait on this thread we'll be blocking rasterization work

If we sleep we also block rasterization. The main purpose of sleep (or busy wait) is to really *waste* some time in the rasterization thread, such that we reduce janks (which looks counter intuitive but see explanations above and experiments in flutter_smooth it seems to work).

> If we sleep we have no way of knowing if the OS will actually reschedule us when we want (just that it probably won't be sooner than that).

Yes I totally agree, that's why I propose busy wait.

Reply:
Ahh sure, but a busy wait for that long will cause the CPU to run hot and use a lot of energy.

Reply:
@dnfield Well I guess indeed not a problem, because:

> For example, suppose most rasterization ends in 15ms and this one ends in 10ms. Then we while(true) noop(); for 5ms.

In such cases, we do busy wait 5ms and seemingly let CPU be wasted (hot and energy) for 5ms. However, most rastierzation needs 15ms, so indeed we just (roughly) pad the 10ms cpu work to 15ms, as if this rasterization job also took 15ms.

Maybe we can use a mixed scheme: Sleep if there is a ton of time left (and wake up a bit earlier) + busy wait for the last bit of time.

In addition, may I know a bit more details about real cases of "sleep needs rescheduling and that happens long after when we want"? I have not seen that in my experiments, and I guess rasterizer thread is of high priority (since it is foreground and we may set it manually).

Another draft idea: we use PostTask a lot, such as start running task on raster thread. But how is PostTask implemented, especially when it gets idle? I do not see it has large latency (as you mentioned in the sleep case), neither should it consume a lot of extra energy (otherwise phones must be very hot since PostTask is used everyday). So maybe we can mimic its implementation?